### PR TITLE
MueLu: add dof-based InterfaceAggregationFactory for Lagrange multiplier aggregates

### LINKED ISSUE #8586: MueLu: extend InterfaceAggregationFactory to dof-based input
## Enhancement
@trilinos/muelu 

This issue shall keep track of efforts to add degree-of-freedom based input to the `MueLu::InterfaceAggregationFactory`.

While PR #8534 results in some non-reproducible errors, let's use this issue to coordinate the splitting into smaller PRs, that are hopefully easier to debug.

## List of branches and PRs

| branch name | PR | status | remarks |
| -------------| :---: | :---: | :---: |
| `muelu-map-transfer-factory` | #8585 | PR closed | replaced by #8593 |
| `muelu-saddle-point-systems` | #8587 | PR closed | PR testing passes, when saddle-point helper is excluded |
| `xpetra-cleanup-and-tests` | #8588 | PR testing queued | PR testing reveals some errors |
| `muelu-coarse-map-factory` | #8589 | PR merged |
| `8586-muelu-lagr-mult-aggregates` | #8593 | PR merged |

### PR DESCRIPTION
**Pull request by mayrmt**
@trilinos/muelu 
@kliegeois 

## Motivation

Add variant of `MueLu::InterfaceAggregationFactory`, that uses a user-given `Xpetra::Map` describing the mortar slave-side interface degrees of freedom of the primal problem. To distinguish between @kliegeois's node-based and the new dof-based variant, the `MueLu::InterfaceAggregationFactory` has a new valid (and mandatory) parameter `"Dual/primal mapping strategy"` with the options `"node-based"` and `"dof-based"`.

To further work with saddle point systems, a `ZeroSubBlockAFactory` is added, which can mimic the extraction of a zero block, which manifests itself as `Teuchos::null` in the input `Xpetra::BlockedCrsMatrix` and as a `Xpetra::Matrix` with proper maps, but no entries on output.

Finally, the `MapTransferFactory` is re-activated and equipped with tests to be used for transferring the interface dof row map (required by the `MueLu::InterfaceAggregationFactory`) to coarse levels.

> **Note for reviewers:** While this PR adds a lot of lines, most of them are for MatrixMarket files. The amount to be reviewed is way smaller than the total amount of touched lines in this PR.

## Related Issues

* Closes #8586
* Replaces #8585 

## Stakeholder Feedback

Required for AMG preconditioning of saddle-point systems arising from mortar discretizations of contact mechanic problems

## Testing

Existing tests pass. New capabilities have been covered by new tests.

## Additional Information

This PR replaces #8585. It contains the same capabilities, but its commit history is more organized and cleaner.

### CODE REVIEW

**Comment by mayrmt** [comment]
@trilinos/muelu I'd appreciate if someone could review this PR. The description provides some context for the proposed changes. I'm happy to provide more information to make thinks clearer, if questions arise. Thanks!

**Comment by mayrmt** [comment]
Could someone from @trilinos/muelu please review this PR? Thanks.

**Comment by mayrmt** [comment]
@kliegeois I was told by @cgcgcg that you're gonna review this PR. I'm aware, that there are lots of changes. I tried to provide background information in the PR description. Let me know, if I can be of any assistance.

**Comment by kliegeois** [comment]
> @kliegeois I was told by @cgcgcg that you're gonna review this PR. I'm aware, that there are lots of changes. I tried to provide background information in the PR description. Let me know, if I can be of any assistance.

@mayrmt, I will try to review it this week and, if possible, at the beginning of the week.

**Review by kliegeois — CHANGES_REQUESTED** [review]
Globally, I think that the PR looks great. Thanks for doing this!

I have a first set of small comments and some of them require some small changes.
I will look again, after those changes, and see if I have additional comments or not.

However, I am not a MueLu developer, I would appreciate if a MueLu developer can approve this PR too.

File: packages/muelu/src/CMakeLists.txt [hunk 1]
```diff
@@ -452,3 +452,4 @@ TRIBITS_ADD_LIBRARY(
 # touch CMakeLists.txt because a new file was created in Utils/ExplicitInstantiation of Utils/ForwardDeclaration
 # touch CMakeLists.txt because a new file was created in Utils/ExplicitInstantiation of Utils/ForwardDeclaration
 # touch CMakeLists.txt because a new file was created in Utils/ExplicitInstantiation of Utils/ForwardDeclaration
+# touch CMakeLists.txt because a new file was created in Utils/ExplicitInstantiation of Utils/ForwardDeclaration
```
[the remark this answers is not in the document]
Comment (mayrmt) [reply packages/muelu/src/CMakeLists.txt:455]: That's an auto-generated change by some script, when adding new factories. This is standard operating procedure to trigger the update of CMake related files. One could also delete one of these lines instead.

File: packages/muelu/src/Headers/MueLu_UseShortNamesScalar.hpp [hunk 2]
```diff
@@ -146,15 +146,15 @@ typedef MueLu::HierarchyFactory<Scalar,LocalOrdinal,GlobalOrdinal,Node> Hierarch
 #ifdef MUELU_HIERARCHYUTILS_SHORT
 typedef MueLu::HierarchyUtils<Scalar,LocalOrdinal,GlobalOrdinal,Node> HierarchyUtils;
 #endif
+#ifdef MUELU_INTERFACEAGGREGATIONFACTORY_SHORT
+typedef MueLu::InterfaceAggregationFactory<Scalar,LocalOrdinal,GlobalOrdinal,Node> InterfaceAggregationFactory;
```
[the remark this answers is not in the document]
Comment (mayrmt) [reply packages/muelu/src/Headers/MueLu_UseShortNamesScalar.hpp:150]: This ordering now is the one, that is generated by the script `gen_classes.sh`. I have manually changes the order in a previous PR to be nicely alphabetical, but naturally the script keeps overwriting this. I'd stick to this new ordering as it is reproducible by running the `gen_classes.sh` script. It's my bad, but this change here basically fixes my screw-up from a previous PR.

File: packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_decl.hpp [hunk 3]
```diff
@@ -139,6 +145,28 @@ class InterfaceAggregationFactory : public SingleLevelFactoryBase
...
+
+private:
+  /*! @brief Build dual aggregates based on a given dual-to-primal node mapping
+   *
+   * @param[in] prefix Prefix for screen output
+   * @param[in/out] currentLevel Level on which the aggregation needs to be performed
+   */
+  void BuildBasedOnNodeMapping(const std::string& prefix, Level& currentLevel) const;
+
+  /*! @brief Build dual aggregates based one given interface row maps of the primal and dual problem
```
[the remark this answers is not in the document]
Comment (mayrmt) [reply packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_decl.hpp]: Done.

File: packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_decl.hpp [hunk 4]
```diff
@@ -139,6 +145,28 @@ class InterfaceAggregationFactory : public SingleLevelFactoryBase
...
+   *
+   * The row map of the interface portion of the primal problem corresponds to the row map of the dual problem.
+   * This correspondence is expoited to form the dual aggregates based on available primal aggregates.
```
[the remark this answers is not in the document]
Comment (mayrmt) [reply packages/muelu/src/Misc/MueLu_InterfaceAggregationFactory_decl.hpp]: Done.

**Comment by mayrmt** [comment]
@kliegeois Thanks for looking into this! I responded to your first set of comments. Let me know if there are more open points to discuss.

Regarding approval, I guess that your opinion is very valuable here since you're more familiar with the `InterfaceAggregationFactory` than other MueLu developers.

**Review by cgcgcg — APPROVED** [review]

**Comment by kliegeois** [comment]
> @kliegeois Thanks for looking into this! I responded to your first set of comments. Let me know if there are more open points to discuss.
> 
> 
> 
> Regarding approval, I guess that your opinion is very valuable here since you're more familiar with the `InterfaceAggregationFactory` than other MueLu developers. 

Thanks @mayrmt !
This looks good to me, thanks for your changes and your explanations.

**Review by kliegeois — APPROVED** [review]

**Pull request merged**